src/masternode-tiers.cpp

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2017-2019 The Transcendence developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "masternode-tiers.h"
+#include <cmath>
+
+bool IsMasternodeOutput(CAmount nValue, int blockHeight)
+{
+    return (GetMasternodeTierFromOutput(nValue, blockHeight) != MasternodeTiers::TIER_NONE);
+}
+
+int GetMasternodeTierFromOutput(CAmount nValue, int blockHeight)
+{
+    int tierRet = MasternodeTiers::TIER_NONE;
+
+    if (blockHeight < TIER_BLOCK_HEIGHT) {
+        if (nValue == MASTERNODE_TIER_COINS[MasternodeTiers::TIER_1K] * COIN) {
+            tierRet = MasternodeTiers::TIER_1K;
+        }
+    }
+    else {
+        for (int tier = MasternodeTiers::TIER_1K; tier != MasternodeTiers::TIER_NONE; tier++) {
+            if (nValue == MASTERNODE_TIER_COINS[tier] * COIN) {
+                tierRet = tier;
+                break;
+            }
+        }
+    }
+    return tierRet;
+}
+
+double GetObfuscationValueForTier(int nTier)
+{
+    if (nTier >= MasternodeTiers::TIER_NONE || nTier < MasternodeTiers::TIER_1K) {
+        return 0;
+    }
+    else {
+        return TIER_OBFUSCATION_VALUES[nTier];
+    }
+}
+
+unsigned int CalculateWinningTier(std::vector<size_t>& vecTierSizes, uint256 blockHash)
+{
+    const unsigned int distribution[MasternodeTiers::TIER_NONE] = {1, 3, 10, 30, 100};
+    double nDenominator = 0; // Summ( distribution[i]*count[i] )
+    unsigned int nMod = 0; // modulus = Summ( distribution[i] )
+
+    //Contains pairs <tier number, weighted value>
+    std::vector<std::pair<size_t, unsigned int>> weightedDistribution;
+
+    //Select tiers which contain nodes
+    for (auto i = 0; i < MasternodeTiers::TIER_NONE; i++) {
+        if (vecTierSizes[i] > 0) {
+            nMod += distribution[i];
+            nDenominator += distribution[i] * vecTierSizes[i];
+            weightedDistribution.push_back(std::make_pair(i, 0));
+        }
+    }
+
+    //Stop calculation if there are no nodes or the only single tier is presented in the network
+    if (weightedDistribution.size() == 0) {
+        return MasternodeTiers::TIER_NONE;
+    }
+    else if (weightedDistribution.size() == 1) {
+        return weightedDistribution[0].first;
+    }
+
+    unsigned int nPreviousWeight = 0;
+    for (auto j = 0; j < weightedDistribution.size() - 1; j++) {
```

Comment:
it seems you can combine two cycles in one. May be all of them in one

Reply:
I do not think we should combine cycles: they represent different logical steps in the algorithm and combining will make it more obfuscated. The first one is for distribution calculation and the second one is tier calculation. Also the first one needs one iteration less.